### PR TITLE
[evm] snapshot gas refund

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
with [this commit](https://github.com/ethereum/go-ethereum/commit/9393d1fb5d178667e4d2750f7f540a3db5985818), the gas refund is changed before a write-opcode bails out in a readOnly tx, we need to snapshot and recover gas refund when reverting the tx

Fixes (#3475)

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [x] fullsync (WIP to verify)
- [] Other test (please specify)

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/execution/evm/evm.go [hunk 1]
```diff
@@ -418,6 +421,18 @@ func executeInEVM(ctx context.Context, evmParams *Params, stateDB *StateDBAdapte
 		// After EIP-3529: refunds are capped to gasUsed / 5
 		refund = (evmParams.gas - remainingGas) / params.RefundQuotientEIP3529
 	}
+	// adjust refund due to dynamicGas
+	var (
+		refundBeforeSnapshot = evm.RefundBeforeSnapshot
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:426]: see https://github.com/iotexproject/go-ethereum/commit/d5ebcb197ff633484324e12cf8f9e08cf262fd56

File: go.mod [hunk 2]
```diff
@@ -201,6 +203,6 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
 
-replace github.com/ethereum/go-ethereum => github.com/iotexproject/go-ethereum v1.7.4-0.20220515212948-4dae4279da68
+replace github.com/ethereum/go-ethereum => github.com/iotexproject/go-ethereum v1.7.4-0.20220913164305-d5ebcb197ff6
```
Comment (dustinxie) [inline go.mod:206]: this will become a tag after fullsync test

File: action/protocol/execution/evm/evm.go [hunk 3]
```diff
@@ -418,6 +421,18 @@ func executeInEVM(ctx context.Context, evmParams *Params, stateDB *StateDBAdapte
...
+		currentRefund        = stateDB.GetRefund()
+	)
+	if !g.IsOkhotsk(blockHeight) && refundBeforeSnapshot > 0 && refundBeforeSnapshot != currentRefund {
+		if refundBeforeSnapshot > currentRefund {
```
Comment (Liuhaai) [inline action/protocol/execution/evm/evm.go:430]: add one more comment before L430?
`// stateDB.refund = refundBeforeSnapshot`

**Review by Liuhaai — APPROVED** [review]

File: action/protocol/execution/evm/evm.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go]: comment is on L424

File: action/protocol/execution/evm/evm.go [hunk 4]
```diff
@@ -317,6 +317,9 @@ func prepareStateDB(ctx context.Context, sm protocol.StateManager) (*StateDBAdap
 	if !featureCtx.FixUnproductiveDelegates {
 		opts = append(opts, NotCheckPutStateErrorOption())
 	}
+	if !featureCtx.CorrectGasRefund {
```
Comment (Liuhaai) [inline action/protocol/execution/evm/evm.go:320]: `if featureCtx.NotCorrectGasRefund {}`

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 5]
```diff
@@ -540,6 +552,13 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
 		log.L().Error("Failed to get snapshot.", zap.Int("snapshot", snapshot))
 		return
 	}
+	// restore gas refund
+	if !stateDB.notCorrectGasRefund {
```
Comment (Liuhaai) [inline action/protocol/execution/evm/evmstatedbadapter.go:556]: if stateDB.notCorrectGasRefund {} 
else {
}

Comment (dustinxie) [reply action/protocol/execution/evm/evmstatedbadapter.go:556]: @Liuhaai  i also feel double negative is confused, i used `correctGasRefund` at the beginning

@CoderZhi can we use `correctGasRefund = true` for the default case (that is after hard-fork)? I know you would like it to be `false` but IMO it's not a big issue (as compared to use double negative and code readability). Let me know

File: action/protocol/execution/evm/evm.go [hunk 6]
```diff
@@ -418,6 +421,18 @@ func executeInEVM(ctx context.Context, evmParams *Params, stateDB *StateDBAdapte
 		// After EIP-3529: refunds are capped to gasUsed / 5
 		refund = (evmParams.gas - remainingGas) / params.RefundQuotientEIP3529
 	}
+	// adjust refund due to dynamicGas
+	var (
+		refundLastSnapshot = stateDB.RefundLastSnapshot()
+		currentRefund      = stateDB.GetRefund()
+	)
+	if evmErr != nil && !g.IsOkhotsk(blockHeight) && refundLastSnapshot > 0 && refundLastSnapshot != currentRefund {
```
Comment (CoderZhi) [inline action/protocol/execution/evm/evm.go:429]: even though, genesis and blockheight are both available here, it is better to use `featureCtx.CorrectGasRefund`.

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 5 above)
Comment (CoderZhi) [reply action/protocol/execution/evm/evmstatedbadapter.go:556]: 1. you misunderstand what @Liuhaai means...
2. `notCorrectGasRefund` could be renamed to `revertRefund`.

File: tools/actioninjector.v2/internal/cmd/inject.go [hunk 7]
```diff
@@ -261,7 +261,7 @@ func (p *injectProcessor) inject(workers *sync.WaitGroup, ticks <-chan uint64) {
 	}
 }
 
-func (p *injectProcessor) pickAction() (iotex.SendActionCaller, error) {
+func (p *injectProcessor) pickAction() (iotex.Caller, error) {
```
Comment (CoderZhi) [inline tools/actioninjector.v2/internal/cmd/inject.go:264]: shall we have a new PR for this unrelated change?

File: api/grpcserver.go [hunk 8]
```diff
@@ -671,7 +672,7 @@ func (svr *gRPCHandler) TraceTransactionStructLogs(ctx context.Context, in *iote
 	if !ok {
 		return nil, status.Error(codes.InvalidArgument, "the type of action is not supported")
 	}
-	tracer := vm.NewStructLogger(nil)
+	tracer := logger.NewStructLogger(nil)
```
Comment (CoderZhi) [inline api/grpcserver.go:675]: do we have to make this change in this PR?

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 9]
```diff
@@ -1010,6 +1036,7 @@ func (stateDB *StateDBAdapter) getNewContract(addr hash.Hash160) (Contract, erro
 
 // clear clears local changes
 func (stateDB *StateDBAdapter) clear() {
+	stateDB.refundSnapshot = make(map[int]uint64)
```
Comment (CoderZhi) [inline action/protocol/execution/evm/evmstatedbadapter.go:1039]: reset `refundLastSnapshot`

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 10]
```diff
@@ -622,6 +641,11 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
 	}
 }
 
+// RefundLastSnapshot returns refund at last snapshot
+func (stateDB *StateDBAdapter) RefundLastSnapshot() uint64 {
```
Comment (CoderZhi) [inline action/protocol/execution/evm/evmstatedbadapter.go:645]: `RefundAtLastSnapshot`

**Review by CoderZhi — APPROVED** [review]

File: api/grpcserver.go (hunk 8 above)
Comment (dustinxie) [reply api/grpcserver.go:675]: yes, due to EVM upgrade

File: action/protocol/execution/evm/evm.go (hunk 6 above)
Comment (Liuhaai) [reply action/protocol/execution/evm/evm.go:429]: same

**Comment by raullenchai** [comment]
Yep!
